crates/krator/src/admission.rs

```diff
@@ -0,0 +1,115 @@
+//! Basic implementation of Kubernetes Admission API
+use crate::Operator;
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::Status;
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+
+/// Result of admission hook.
+#[allow(clippy::large_enum_variant)]
+pub enum AdmissionResult<T> {
+    /// Permit the request. Pass the object (with possible mutations) back.
+    /// JSON Patch of any changes will automatically be created.
+    Allow(T),
+    /// Deny the request. Pass a Status object to provide information about the error.
+    Deny(Status),
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+/// AdmissionRequest describes the admission.Attributes for the admission request.
+struct AdmissionRequest<T> {
+    /// UID is an identifier for the individual request/response. It allows us to distinguish instances of requests which are
+    /// otherwise identical (parallel requests, requests when earlier requests did not modify etc)
+    /// The UID is meant to track the round trip (request/response) between the KAS and the WebHook, not the user request.
+    /// It is suitable for correlating log entries between the webhook and apiserver, for either auditing or debugging.
+    uid: Option<String>,
+    /// Object is the object from the incoming request.
+    object: T,
+}
+
+/// AdmissionResponse describes an admission response.
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct AdmissionResponse {
+    /// UID is an identifier for the individual request/response.
+    /// This must be copied over from the corresponding AdmissionRequest.
+    uid: Option<String>,
+    /// Allowed indicates whether or not the admission request was permitted.
+    allowed: bool,
+    /// Result contains extra details into why an admission request was denied.
+    /// This field IS NOT consulted in any way if "Allowed" is "true".
+    status: Option<Status>,
+    /// The patch body. Currently we only support "JSONPatch" which implements RFC 6902.
+    patch: Option<json_patch::Patch>,
+    /// The type of Patch. Currently we only allow "JSONPatch".
+    patch_type: Option<String>,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct AdmissionReviewRequest<T> {
```

Comment:
Interesting, does `T` not need to be `Deserialize`?

Reply:
I think if you tried to use a `T` that isn't `Deserialize`, the `derive` would fail to compile. But this is a private type that is only used in `endpoint` so I don't think this is a concern.